### DISCUSSION TITLE
Cannot get CF position when flying more than 3 CF

### DISCUSSION THREAD

**Discussion by Dhipz** [USC-ACTLab/crazyswarm#698] [Q&A]
Hi, so i'm trying to get the CF position (x, y, z, coordinates) while flying and i was able to do it for up to 3 CF flying simultaneously. But every time i tried using more than 3 CF only 3 of the CF showed up on the rviz and the code i'm running always gives this errors.

```
/home/user/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py:590: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  cfg = yaml.load(crazyflies_yaml)
Warning: No joystick found!
Traceback (most recent call last):
  File "code.py", line 16, in <module>
    c_p = cf.position()
  File "/home/user/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py", line 337, in position
    self.tf.waitForTransform("/world", "/cf" + str(self.id), rospy.Time(0), rospy.Duration(10))
  File "/opt/ros/noetic/lib/python3/dist-packages/tf/listener.py", line 76, in waitForTransform
    raise tf2_ros.TransformException(error_msg or "no such transformation: \"{}\" -> \"{}\"".format(source_frame, target_frame))
tf2.TransformException: canTransform: source_frame cf3 does not exist.
```

This is the test code i used
```py
#!/usr/bin/env python

import numpy as np

from pycrazyswarm import *

swarm = Crazyswarm()
timeHelper = swarm.timeHelper
allcfs = swarm.allcfs


for cf in allcfs.crazyflies:
    cf.takeoff(targetHeight=1.0, duration=2.0)
timeHelper.sleep(3.0)
for cf in allcfs.crazyflies:
    c_p = cf.position()
    print("ID = ", cf.id, " | POS = ", c_p)
    curr_pos = [c_p[0]-0.5,c_p[1],c_p[2]]
    cf.goTo(curr_pos,0,2)
timeHelper.sleep(2.5)
allcfs.land(targetHeight=0.06, duration=2.0)
timeHelper.sleep(3.0)
```

My setup used 2 Lighthouse V2, default CF with lighthouse deck, and all the launch settings are default with the logging enabled and `motion_capture_type` set to none

**Comment by Dhipz** [comment]
@whoenig @jpreiss I'm sorry for tagging but please take a look when possible. Thank you.

**Comment by whoenig** [comment]
The first thing to double check would be if the /pose topics contain the relevant data. My guess is that they are not and that the core of the issue is based on streaming logging data (a use-case Crazyswarm hasn't been optimized/tested for).

To improve the situation:

- Decrease the waiting time at https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L971
- Play with the update rate at https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L609

You might also want to try Crazyswarm2 (if you are ready for ROS2); this has one backend based on the official Python library, which might behave better in your use-case.

**Reply by Dhipz** [reply]
Thank you for the reply! Sorry i'm not familiar with ROS, could you explain a tiny bit about what both of the solutions originally do? 
And i'll definitely check out Crazyswarm2! Is there any sources i could read about it, especially about the difference between it and Crazyswarm1?

Once again, thank you!

**Reply by Dhipz** [reply]
Hi so i've tried playing with both of the values but it didn't improve the logging, there's still only 3 CF shown in the rviz. I also checked the /pose topics and yeah, there is no data when i tried to echo it for any CF that are not in the rviz. What i eventually did is using 3 separate channels for 6 CF. Although i find it weird that only 4 of the CF shows up in rviz when i used 2 channels while 3 CF shows up when i used only 1 channel. Anyway i hope Crazyswarm2 with the official backend would fix this issue. Thank you.

**Reply by AvaneenP** [reply]
@Dhipz Hi did you ever have this problem when working with just two CFs? I'm having a very similar problem where I only get the frame for one CF when I launch hover_swarm.launch. Here's a video that shows that: https://drive.google.com/file/d/1ArSWtq4aDP4r80F_VT91RwUND2fSa9Pp/view?usp=sharing

And did you resolve it in ROS1 or did you just end up going to ROS2?
